Change timeout kill signals to SIGQUIT from SIGTERM

staticfloat opened the pull request:
We've been having a lot of timeouts on CI recently.  Our coredumps might
be helpful in tracking these down, but when we send `SIGTERM` or
`SIGKILL` we don't get coredumps.  This allows customization of which
signal is sent during these timeout messages, to allow for CI to set it
to `SIGSEGV` or similar, to force core dumps for debugging purposes.

DilumAluthge reviewed (approved):
I think we have a convention in this repo to do `JULIA_TEST_*` instead of `JULIA_TESTING_*`, so it would be good to be consistent.

DilumAluthge commented on stdlib/Distributed/src/managers.jl:
```diff
@@ -736,7 +736,10 @@ function kill(manager::LocalManager, pid::Int, config::WorkerConfig; exit_timeou
             @warn("Failed to gracefully kill worker $(pid), sending SIGTERM")
-            kill(config.process, Base.SIGTERM)
+
+            # Support overriding this for the 
```
This comment looks incomplete?

staticfloat commented:
Canceling buildkite build, since it's truly being tested here: https://github.com/JuliaCI/julia-buildkite/pull/165

DilumAluthge commented:
@vtjnash Could you explain why this is a bad idea?

vtjnash commented:
Seemed somewhat awkward for the API, but I didn't really think much about it so I didn't leave a review. It also may mean we won't generate a backtrace before exiting anymore

Unrelatedly: currently this appears to be also killing much too fast, so our CI logs are getting incorrectly littered with errors from this, and leaving behind some tmp files that seemed to sometimes now cause some issues for the machines

DilumAluthge commented:
@staticfloat Can you rebase and fix the merge conflicts?

staticfloat committed:
Change timeout kill signals to SIGQUIT from SIGTERM

We've been having a lot of timeouts on CI recently.  Our coredumps might
be helpful in tracking these down, but when we send `SIGTERM` or
`SIGKILL` we don't get coredumps.  This changes timeout signals to
generally use `SIGQUIT` instead of `SIGTERM`, when sending the first kill signal.

staticfloat commented:
@vtjnash now that https://github.com/JuliaLang/julia/pull/47459 has made it easier for us to get coredumps on SIGQUIT, I think this PR makes even more sense.  We would define `JULIA_TEST_TIMEOUT_SIGNUM=3` (for `SIGQUIT`) and get core dumps whenever we hit a timeout in the test suite.  Do you agree?

staticfloat commented:
After talking it over with Jameson, he recommended that we skip the whole customization thing and just unconditionally use `SIGQUIT`.

DilumAluthge reviewed (commented):
Why are you hard-coding SIGQUIT, instead of reading it from an environment variable?

DilumAluthge commented:
😂 Left my comment before I saw your comment.

DilumAluthge reviewed (approved):
SGTM

staticfloat commented:
Jameson tells me that one complication with this is that we may not be able to use SIGQUIT on Windows.

DilumAluthge commented:
Use SIGTERM on Windows, and SIGQUIT otherwise?

DilumAluthge commented:
How does one even trigger a core dump on Windows?

vtjnash commented:
There are a couple library functions for that I believe (notably ReportFault) https://learn.microsoft.com/en-us/windows/win32/dxtecharts/crash-dump-analysis

DilumAluthge commented:
Ah, that's right... IIRC Keno had some code in `julia-buildbot` for getting minidumps on the Windows Buildbots. We should port that over to Buildkite.

I don't think that needs to block this PR though.

staticfloat commented:
So Jameson what do you suggest here?  Should we embed `Sys.iswindows() ? SIGTERM : SIGQUIT` everywhere?  Should we define `Base.SIGQUIT = Base.SIGTERM`?  Should we add a patch to libuv to support `SIGQUIT` in the same way it supports `SIGTERM`?

vtjnash commented:
I think we can patch libuv to make this fatal, but currently to define SIGQUIT=SIGTERM also

vtjnash commented:
For extra bonus points, you could even make a dump file when sending SIGQUIT (since there is an API function for doing just that)

staticfloat commented:
Here's my initial stab at this: https://github.com/JuliaLang/libuv/pull/30

staticfloat commented:
Switching this PR to draft until the following three PRs are merged:

- https://github.com/JuliaLang/libuv/pull/30
- https://github.com/JuliaLang/libuv/pull/31
- https://github.com/JuliaLang/julia/pull/47611

staticfloat commented:
Okay, I rebased and moved the `Distributed` changes over to here: https://github.com/JuliaLang/Distributed.jl/pull/103
